win32ss/gdi/ntgdi/freetype.c

```diff
@@ -1577,16 +1601,39 @@ IntGdiAddFontResourceEx(PUNICODE_STRING FileName, DWORD Characteristics,
     PVOID Buffer = NULL;
     IO_STATUS_BLOCK Iosb;
     PVOID SectionObject;
-    SIZE_T ViewSize = 0;
+    SIZE_T ViewSize = 0, Length;
     LARGE_INTEGER SectionSize;
     OBJECT_ATTRIBUTES ObjectAttributes;
     GDI_LOAD_FONT LoadFont;
     INT FontCount;
     HANDLE KeyHandle;
+    UNICODE_STRING PathName;
+    LPWSTR pszBuffer;
     static const UNICODE_STRING TrueTypePostfix = RTL_CONSTANT_STRING(L" (TrueType)");
+    static const UNICODE_STRING DosPathPrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
+
+    /* Build PathName */
+    if (dwFlags & AFRX_DOS_DEVICE_PATH)
+    {
+        Length = DosPathPrefix.Length + FileName->Length + sizeof(UNICODE_NULL);
+        pszBuffer = ExAllocatePoolWithTag(PagedPool, Length, TAG_USTR);
+        if (!pszBuffer)
+            return 0;   /* failure */
+
+        RtlInitEmptyUnicodeString(&PathName, pszBuffer, Length);
+        RtlAppendUnicodeStringToString(&PathName, &DosPathPrefix);
+        RtlAppendUnicodeStringToString(&PathName, FileName);
+    }
+    else
+    {
+        Status = DuplicateUnicodeString(FileName, &PathName);
+        if (!NT_SUCCESS(Status))
+            return 0;   /* failure */
+    }
 
     /* Open the font file */
-    InitializeObjectAttributes(&ObjectAttributes, FileName, 0, NULL, NULL);
+    InitializeObjectAttributes(&ObjectAttributes, &PathName,
+                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE, NULL, NULL);
```

Comment:
This `OBJ_KERNEL_HANDLE` is what makes `MmCreateSection` below fail with `STATUS_INVALID_HANDLE`

Reply:
Deleted `OBJ_KERNEL_HANDLE` flag.

Reply:
@ThFabba How do you think this flag?
```txt
(ntoskrnl/mm/section.c:5039) Failed to get a handle to the FO: c0000008
```

Reply:
Yeah it looks like `MmCreateSection` always references the `FileHandle` using `ExGetPreviousMode()`, which will be UserMode for most of win32k.
You could either `ObReferenceObjectByHandle` the file handle here and pass the `FILE_OBJECT` to MmCreateSection (though there seem to be slighty behavior differences in there for handle vs object), or use `ZwCreateSection` instead and then `ObReferenceObjectByHandle` to get the section object from the section handle.

Reply:
> This `OBJ_KERNEL_HANDLE` is what makes `MmCreateSection` below fail with `STATUS_INVALID_HANDLE`
So, is it an actual bug in ROS code (i.e. does not happen on Windows), or is it an actual feature?